proposals/lip-0039.md

```diff
@@ -302,273 +328,365 @@ proof={siblingHashes:[cc956a85, 3c516152, e041e1c0, 6400721e, 3c32b131],
+            return False
+
+    return verify(query_keys, proof, root)
 ```
 
-```java
-// This function checks whether two queries correspond to nodes that are
-// children of the same branch node, with q1 and q2 the left and right
-// child respectively
-function areSiblingQueries(q1, q2):
-    if length of q1.binaryBitmap != length of q2.binaryBitmap:
-        return false
-    let h=length of q1.binaryBitmap
-    let binaryKey1=binaryExpansion(q1.key)
-    let binaryKey2=binaryExpansion(q2.key)
-    let keyPrefix1 be the first (h-1) digits of binaryKey1
-    let keyPrefix2 be the first (h-1) digits of binaryKey2
-    if keyPrefix1 != keyPrefix2:
-        return false
-    let d1 be the digit at position h of binaryKey1
-    let d2 be the digit at position h of binaryKey2
-    return d1==0 and d2==1
+```python
+def verify_non_inclusion_proof(query_keys: list[bytes], proof: Proof, root: bytes) -> bool:
+    # Check that there is a query for each key.
+    if len(query_keys) != len(proof.queries):
+        return False
+
+    for key, query in zip(query_keys, proof.queries):
+        # This is the only difference between an inclusion and a non-inclusion proof.
+        if is_inclusion_proof(key, query) == True:
+            return False
+            
+    return verify(query_keys, proof, root)
 ```
 
-```java
-function calculateRoot(siblingHashes, queries):
-    // we add the hash property to each query, used to recompute the tree root
-    for each query q in queries:
-        set q.binaryBitmap to binaryExpansion(q.bitmap) without leading zeros
-        // if value is empty, it is a defaultEmptyNode
-        if q.value is empty bytes:
-            set q.hash to EMPTY_HASH
+```python
+def verify(query_keys: list[bytes], proof: Proof, root: bytes) -> bool:
+    for key, query in zip(query_keys, proof.queries):
+        # The bitmap does not contain extra leading 0s.
+        if query.bitmap[0] == 0:
+            return False
+        if len(key) != KEY_LENGTH_BYTES:
+            return False
+
+        # q is an inclusion proof for k or a default empty node.
+        if key == query.key:
+            continue
+
+        # q is an inclusion proof for another leaf node
+        common_prefix = os.path.commonprefix([binary_expansion(query.key), binary_expansion(key)])
+        if len(query.binary_bitmap) > len(common_prefix):
+            # q does not give an non-inclusion proof for k.
+            return False
+
+    filtered_queries = []
+    filter: dict[bytes, bool] = {}
+    duplicate_queries = {}
+    for q in proof.queries:
+        # Remove duplicate queries preserving order.
+        # This can happen if the same query is given for different query_keys, as for inclusion proofs or non-inclusion proofs pointing to a leaf node
+        # or for non-inclusion proofs pointing to the same empty node.
+        # To do this, we check the binary path (key binary expansion up to the query height)
+        if q.binary_path not in filter:
```

Comment:
I don't understand why unique q.binary_path and not unique q.key. Maybe we can discuss in person.

Reply:
updated after discussion, now we check that duplicate queries (queries with equal binary path) point to exactly the same node

Reply:
@shuse2 We updated the LIP to fix a possible vulnerability. Can you double check how this is done in the DB?

Reply:
the verification in the implementation is done [here](https://github.com/LiskHQ/lisk-db/blob/main/src/sparse_merkle_tree/smt.rs#L845-L877), which is used in https://github.com/LiskHQ/lisk-db/blob/main/src/sparse_merkle_tree/smt.rs#L1528

Reply:
Can be resolved from my side.